### PR TITLE
Images default architecture

### PR DESCRIPTION
**Pull request by presztak**
Add concept of default architecture (global and per-project) #7947

### CODE REVIEW

File: lxd/api_project.go [hunk 1]
```diff
@@ -764,14 +765,17 @@ func isEitherAllowOrBlockOrManaged(value string) error {
...
+		"features.profiles":             validate.Optional(validate.IsBool),
+		"features.images":               validate.Optional(validate.IsBool),
+		"features.storage.volumes":      validate.Optional(validate.IsBool),
+		"features.networks":             validate.Optional(validate.IsBool),
+		"images.auto_update_cached":     validate.Optional(validate.IsBool),
+		"images.auto_update_interval":   validate.Optional(validate.IsInt64),
+		"images.compression_algorithm":  validate.IsCompressionAlgorithm,
+		"images.default_architecture": func(value string) error {
+			return validate.IsOneOf(value, osarch.SupportedArchitectures())
+		},
```
Comment (stgraber) [inline lxd/api_project.go:778]: Could probably define a `validate.IsArchitecture` function to make this even easier to reuse elsewhere.

File: lxd/cluster/config.go [hunk 2]
```diff
@@ -246,6 +247,7 @@ var ConfigSchema = config.Schema{
 	"images.auto_update_cached":      {Type: config.Bool, Default: "true"},
 	"images.auto_update_interval":    {Type: config.Int64, Default: "6"},
 	"images.compression_algorithm":   {Default: "gzip", Validator: validate.IsCompressionAlgorithm},
+	"images.default_architecture":    {Validator: defaultArchValidator},
```
Comment (stgraber) [inline lxd/cluster/config.go:250]: With the step mentioned in the previous commit, you could then use `validate.IsArchitecture` here.

**Comment by stgraber** [comment]
Looking really good, thanks!

I'll need to take another closer look at `GetNodeWithLeastInstances` when I'm less tired (off today/tomorrow so just doing some reviews from my phone here and there). And I think the addition of `validate.IsArchitecture` will make the rest even tidier (not to mention useful for other places where we deal with architectures).

File: lxd/api_project.go (hunk 1 above)
Comment (presztak) [reply lxd/api_project.go:778]: @stgraber Yes, this is how I wanted to do this but after import `osarch` package in `validate.go` I get error related with cycle import:
```
import cycle not allowed
package github.com/lxc/lxd/client
	imports github.com/lxc/lxd/shared
	imports github.com/lxc/lxd/shared/validate
	imports github.com/lxc/lxd/shared/osarch
	imports github.com/lxc/lxd/shared
```

so this is why I made this in this way.

File: lxd/cluster/config.go (hunk 2 above)
Comment (presztak) [reply lxd/cluster/config.go:250]: @stgraber same problem here as in my previous answer.

File: lxd/api_project.go (hunk 1 above)
Comment (stgraber) [reply lxd/api_project.go:778]: Ah, that's slightly annoying :)

Comment (stgraber) [reply lxd/api_project.go]: Fixed that for you now.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
@stgraber happy for me to merge this?

**Comment by stgraber** [comment]
@tomponline yep

**Pull request merged**